app/herdbook.py

```diff
@@ -505,6 +509,314 @@ def testbreed():
     return {"offspringCOI": formatted_offspring_coi}
 
 
+@APP.route("/api/certificates/update/<i_number>", methods=["PATCH"])
+@login_required
+def update_certificate(i_number):
+    """
+    Returns an updated pdf of the individual given by `i_number`.
+    """
+    user_id = session.get("user_id", None)
+    ind = da.get_individual(i_number, user_id)
+
+    if ind is None:
+        return jsonify({"response": "Individual not found"}), 404
+
+    certificate_exists = ind.get("certificate", None)
+    form = request.json
+    uploaded = False
+
+    try:
+        present = check_certificate_s3(ind_number=ind["number"])
+        if certificate_exists and present:
+            data = get_certificate_data(ind, user_id)
+            data.update(**form)
```

Comment:
I still don't like this and don't think it does what's intended (at least not always).

Reply:
I still don't understand what you want either. Can you please explain?

Reply:
Hrrm, misplaced this comment, should be on the `ind.update` line.

For that; what it looks like it should do is start creating an update query (http://docs.peewee-orm.com/en/latest/peewee/api.html#Model.update which doesn't change anything). Then call the `update_individual` to do update (which is intended to update from a form which is not exactly what we're doing here).

So again, I don't understand how the ind.update can change anything (given as for my current understanding, it just creates an update query that is not executed but immediately forgotten). 

And if it does work as intended, I still don't like it on account on being confusing.

I'd much prefer having a whitelist of allowed changes here (keys in dict `form`) and passing to a new function in `data_access` to update individual according to that form. That could either do a

```
    ind = get_individual(i_number, user_id)
    ind.update(**form).where(Individual.number == i_number).execute()
```   
or
```
    ind = get_individual(i_number, user_id)
    for k,v in form:
        ind.setattr(k,v)
    ind.save()
```   

(both inside `with DATABASE.atomic():`)

Reply:
See https://github.com/NBISweden/theherdbook/commit/ef1bfbe4aa910ba6e6d7fe4ac650fda8db325ab8  for some minumum changes (only naming and comments) to lessen confusion. I'd prefer other changes but that would bloat this PR even more, so it's something for the future.